agent/src/agent_win.cc

```diff
@@ -11,87 +11,383 @@
+  }
+
+  is_connected_ = false;
+  browser_info_ = BrowserInfo();
+  buffer_.clear();
+  cursor_ = nullptr;
+  read_size_ = 0;
+  final_size_ = 0;
+
+  // Cancel any outstanding IO requests on this pipe.
+  if (handle_ != INVALID_HANDLE_VALUE) {
+    CancelIoEx(handle_, nullptr);
+  }
+
+  // This function does not close `handle_` or the event in overlapped so
+  // that the server can resuse the pipe with an new Google Chrome browser
+  // instance.
+}
+
+DWORD AgentWin::Connection::QueueReadFile(bool reset_cursor) {
+  if (reset_cursor) {
+    cursor_ = buffer_.data();
+    read_size_ = buffer_.size();
+    final_size_ = 0;
+  }
+
+  // When this function is called there are the following possiblities:
+  //
+  // 1/ Data is already available and the buffer is filled in.  ReadFile()
+  //    return TRUE and the event is set.
+  // 2/ Data is not avaiable yet.  ReadFile() returns FALSE and the last error
+  //    is ERROR_IO_PENDING(997) and the event is reset.
+  // 3/ Some error occurred, like for example Google Chrome stops.  ReadFile()
+  //    returns FALSE and the last error is something other than
+  //    ERROR_IO_PENDING, for example ERROR_BROKEN_PIPE(109).  The event
+  //    state is unchanged.
+  DWORD err = ERROR_SUCCESS;
+  DWORD count;
+  if (!ReadFile(handle_, cursor_, read_size_, &count, &overlapped_)) {
+    err = GetLastError();
+  }
+
+  return err;
+}
+
+DWORD AgentWin::Connection::OnReadFile(BOOL success, DWORD count) {
+  final_size_ += count;
+
+  // If success is TRUE, this means the full message has been read.  Call the
+  // appropriate handler method.
+  if (success) {
+    return CallHandler();
+  }
+
+  // If success os false, there are two possibilities:
+  //
+  // 1/ The last error is ERROR_MORE_DATA(234).  This means there are more
+  //     bytes to read before the request message is complete.  Resize the
+  //     buffer and adjust the cursor.  The caller will queue up another read
+  //     and wait.
+  // 2/ Some error occured.  In this case return the error.
+
+  DWORD err = GetLastError();
+  if (err == ERROR_MORE_DATA) {
+    read_size_ = kBufferSize;
+    buffer_.resize(buffer_.size() + read_size_);
+    cursor_ = buffer_.data() + buffer_.size() - read_size_;
+  }
+
+  return err;
+}
+
+DWORD AgentWin::Connection::CallHandler() {
+  DWORD err = ERROR_SUCCESS;
+
+  ChromeToAgent request;
+  if (request.ParseFromArray(buffer_.data(), final_size_)) {
+    if (request.has_request()) {
+      auto event = std::make_unique<ContentAnalysisEventWin>(
+          handle_, browser_info_, std::move(*request.mutable_request()));
+      if (event->Init() == ERROR_SUCCESS) {
+        handler_->OnAnalysisRequested(std::move(event));
+      } else {
+        err = ERROR_INTERNAL_ERROR;
+      }
+    } else if (request.has_ack()) {
+      handler_->OnResponseAcknowledged(request.ack());
+    } else {
+      // Malformed request.
+      err = ERROR_INVALID_DATA;
+    }
+  }
```

Comment:
Same here, would be nice to have comment for each scenario and/or reduce nested if.